kore/src/Kore/Step/SMT/AST.hs

```diff
@@ -0,0 +1,178 @@
+
+data KoreSortDeclaration sort symbol name
+    = SortDeclarationDataType !(AST.DataTypeDeclaration sort symbol name)
+    | SortDeclarationSort !(AST.SortDeclaration name)
+    | SortDeclaredIndirectly !name
+    deriving (Eq, Ord, Show)
+
+data KoreSymbolDeclaration sort name
+    = SymbolDeclaredDirectly !(AST.FunctionDeclaration sort name)
+    | SymbolDeclaredIndirectly !name ![sort]
+    deriving (Eq, Ord, Show)
+
+data Declarations sort symbol name =
+    Declarations
+        { sorts :: Map (Kore.Id Object) (Sort sort symbol name)
+        , symbols :: Map (Kore.Id Object) (Symbol sort name)
+        }
+    deriving Show
+
+
+newtype SortReference = SortReference { getSortReference :: Kore.Sort Object }
+    deriving (Eq, Ord, Show)
+newtype SymbolReference =
+    SymbolReference { getSymbolReference :: Kore.Id Object }
+    deriving (Eq, Ord, Show)
+data Encodable
+    = AlreadyEncoded !Text
+    | Encodable !Text
+    -- TODO (virgil): maybe use Id in Encodable to make it more obvious what
+    -- happens.
+    deriving (Eq, Ord, Show)
+
+type SmtDeclarations = Declarations AST.SExpr Text Text
+type SmtKoreSymbolDeclaration = KoreSymbolDeclaration AST.SExpr Text
+type SmtSort = Sort AST.SExpr Text Text
+type SmtSymbol = Symbol AST.SExpr Text
+
+type UnresolvedConstructorArgument =
+    AST.ConstructorArgument SortReference Encodable
+type UnresolvedConstructor =
+    AST.Constructor SortReference SymbolReference Encodable
+type UnresolvedDataTypeDeclaration =
+    AST.DataTypeDeclaration SortReference SymbolReference Encodable
+type UnresolvedDeclarations =
+    Declarations SortReference SymbolReference Encodable
+type UnresolvedFunctionDeclaration =
+    AST.FunctionDeclaration SortReference Encodable
+type UnresolvedKoreSortDeclaration =
+    KoreSortDeclaration SortReference SymbolReference Encodable
+type UnresolvedKoreSymbolDeclaration =
+    KoreSymbolDeclaration SortReference Encodable
+type UnresolvedSort =
+    Sort SortReference SymbolReference Encodable
+type UnresolvedSortDeclaration =
+    AST.SortDeclaration Encodable
+type UnresolvedSymbol =
+    Symbol SortReference Encodable
+
+encodable :: Kore.Id Object -> Encodable
+encodable Kore.Id {getId} = Encodable getId
+
+encode :: Encodable -> Text
+encode (AlreadyEncoded e) = e
+encode (Encodable e) = encodeName e
+
+appendToEncoding :: Encodable -> Text -> Encodable
+appendToEncoding (AlreadyEncoded e) t = AlreadyEncoded (e <> t)
+appendToEncoding (Encodable e) t = Encodable (e <> t)
+
+mergePreferFirst
+    :: Declarations sort symbol name
+    -> Declarations sort symbol name
+    -> Declarations sort symbol name
+mergePreferFirst
+    Declarations { sorts = sorts1, symbols = symbols1 }
+    Declarations { sorts = sorts2, symbols = symbols2 }
+  =
+    Declarations
+        { sorts = sorts1 `leftMerge` sorts2
+        , symbols = symbols1 `leftMerge` symbols2
+        }
+  where
+    leftMerge :: Ord k => Map k v -> Map k v -> Map k v
+    leftMerge =
```

Comment:
The ordinary `Map.union` is also left-biased.

Reply:
Thank you!